torch/_export/non_strict_utils.py

```diff
@@ -0,0 +1,145 @@
+import inspect
+from collections import defaultdict
+from typing import Dict, List
+
+import torch
+from torch._dynamo.source import (
+    GetItemSource,
+    LocalSource,
+    TensorProperty,
+    TensorPropertySource,
+)
+from torch._dynamo.variables.builder import TrackedFake
+from torch._export.passes.add_runtime_assertions_for_constraints_pass import InputDim
+from torch._guards import Source
+from torch._subclasses.fake_tensor import FakeTensorMode
+from torch.export import Constraint
+from torch.fx.experimental.symbolic_shapes import (
+    DimDynamic,
+    EqualityConstraint,
+    ShapeEnv,
+    StatelessSymbolicContext,
+)
+
+
+def fakify(mode, t, tensor_constraints, source, dim_name_to_sources):
+    n_dims = len(t.shape)
+    symbolic_context = StatelessSymbolicContext(
+        dynamic_sizes=[DimDynamic.STATIC] * n_dims,
+        constraint_sizes=[None] * n_dims,
+    )
+    t_id = id(t)
+    if t_id in tensor_constraints:
+        for i, constraint in tensor_constraints[t_id].items():
+            symbolic_context.constraint_sizes[i] = constraint.constraint_range
+            symbolic_context.dynamic_sizes[i] = DimDynamic.DYNAMIC
+            src = TensorPropertySource(base=source, prop=TensorProperty.SIZE, idx=i)
+            dim_name_to_sources[constraint.debug_name].append(src)
+            mode.shape_env.source_name_to_debug_name[src.name()] = constraint.debug_name
+    fake = mode.from_tensor(t, source=source, symbolic_context=symbolic_context)
+    mode.shape_env.tracked_fakes.append(
+        TrackedFake(fake, source, symbolic_context.constraint_sizes)
+    )
+    return fake
+
+
+def fake_tree(mode, arg, tensor_constraints, source, dim_name_to_sources):
+    if isinstance(arg, list):
+        return [
+            fake_tree(
+                mode,
+                arg,
+                tensor_constraints,
+                GetItemSource(source, i),
+                dim_name_to_sources,
+            )
+            for i, arg in enumerate(arg)
+        ]
+
+    elif isinstance(arg, dict):
+        return {
+            k: fake_tree(
+                mode,
+                arg,
+                tensor_constraints,
+                GetItemSource(source, k),
+                dim_name_to_sources,
+            )
+            for k, arg in arg.items()
+        }
+
+    else:
+        return fakify(mode, arg, tensor_constraints, source, dim_name_to_sources)
+
+
+def make_fake_inputs(nn_module, args, constraints):
+    tensor_constraints: Dict[int, Dict[int, Constraint]] = defaultdict(dict)
+    for constraint in constraints:
+        tensor_constraints[constraint.t_id][constraint.dim] = constraint
+        if constraint.shared is not None:
+            tensor_constraints[constraint.shared.t_id][constraint.shared.dim] = constraint
+
+    code = nn_module.forward.__code__
+    co_fields = {
```

Comment:
Just curious: why do we need these in the shape_env

Reply:
I wanted the shape env to be as much the same as what's produced by strict, and I noticed that this is populated in strict mode.